servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Single.java

```diff
@@ -793,7 +793,37 @@ public final Publisher<T> concat(Publisher<? extends T> next) {
      * all elements from {@code next} {@link Publisher}.
      */
     public final Publisher<T> concat(Publisher<? extends T> next, boolean deferSubscribe) {
-        return new SingleConcatWithPublisher<>(this, next, deferSubscribe);
+        return concat(next, deferSubscribe, false);
+    }
+
+    /**
+     * Returns a {@link Publisher} that first emits the result of this {@link Single} and then subscribes and emits all
+     * elements from {@code next} {@link Publisher}. Any error emitted by this {@link Single} or {@code next}
+     * {@link Publisher} is forwarded to the returned {@link Publisher}.
+     * <p>
+     * This method provides a means to sequence the execution of two asynchronous sources and in sequential programming
+     * is similar to:
+     * <pre>{@code
+     *     List<T> results = new ...;
+     *     results.add(resultOfThisSingle());
+     *     results.addAll(nextStream());
+     *     return results;
+     * }</pre>
+     * @param next {@link Publisher} to concat.
+     * @param deferSubscribe if {@code true} subscribe to the {@code next} {@link Publisher} will be deferred until
+     * demand is received. Otherwise, it subscribes to the {@code next} {@link Publisher} as soon as {@code this}
+     * {@link Single} completes successfully. Choosing the deferred ({@code true}) behavior is important if the
+     * {@code next} {@link Publisher} does not or might not support multiple subscribers (non-replayable). Choosing the
+     * immediate subscribe ({@code false}) behavior may have better performance and may be a preferable choice for
+     * replayable {@link Publisher}(s) or when eager subscribe is beneficial.
+     * @param propagateCancel if {@code true} then {@code next} will be subscribed to and cancelled if this
+     * {@link Publisher} is cancelled or terminates with {@link Subscriber#onError(Throwable)}. if {@code false} this
+     * method is equivalent to {@link #concat(Publisher, boolean)}.
+     * @return New {@link Publisher} that first emits the result of this {@link Single} and then subscribes and emits
+     * all elements from {@code next} {@link Publisher}.
+     */
+    public final Publisher<T> concat(Publisher<? extends T> next, boolean deferSubscribe, boolean propagateCancel) {
```

Comment:
The `deferSubscribe` and `propagateCancel` look conflicting. If users use `deferSubscribe == true` they do not expect subscribe to the `next` without demand and success for the `Single`. Javadoc for `deferSubscribe` describes a use-case with non-replayable publisher. Do you have a legit use-case in mind for `deferSubscribe == true && propagateCancel == true`?

My expectation was that we will modify only `deferSubscribe == false` variant to make it consistent with all other `concat` variants: it should subscribe to the `next` after `onSuccess` even if `cancel` was already observed. Because for our use-case we use `Single.succeeded`, it will be enough. No need for extra complexity.

Reply:
These two options are independent and augment the operator behavior in a non-mutually exclusive way. `deferSubscribe` applies to the happy path, `propagateCancel` applies to the not-happy path. In the specific call site we have today in https://github.com/apple/servicetalk/pull/2369 we are using `Single.succeeded` but the semantics we desire from `concat` are provided by `propagateCancel`, so I think it is safer to provide these semantics in the operator rather than rely upon the `Single` implementation (e.g. what if we change it to respect cancel in the future). This also allows us to provide consistent semantics with https://github.com/apple/servicetalk/pull/2372 which is required at our current call-site.

Reply:
discussed offline, we can do both and I will rename operator to `concatPropagateCancel`.